File: backend/services/moderationService.js
import axios from "axios";
import dotenv from "dotenv";

dotenv.config();

const MODERATION_API_URL = process.env.MODERATION_API_URL || "http://localhost:8000/generate/";
const MODERATION_TIMEOUT_MS = parseInt(process.env.MODERATION_TIMEOUT_MS) || 30000;
const MAX_RETRIES = 2;
const RETRY_DELAY = 1000; 

// In-memory rate limiting
let lastRequestTime = 0;
const MIN_REQUEST_INTERVAL = 2000; 

/**
 * Wait fo rate limits
 */
async function waitForRateLimit() {
  const now = Date.now();
  const timeSinceLastRequest = now - lastRequestTime;
  
  if (timeSinceLastRequest < MIN_REQUEST_INTERVAL) {
    const waitTime = MIN_REQUEST_INTERVAL - timeSinceLastRequest;
    console.log(`⏱ Rate limit: Waiting ${waitTime}ms before next request...`);
    await new Promise(resolve => setTimeout(resolve, waitTime));
  }
  
  lastRequestTime = Date.now();
}

/**
 * Sleep helper
 */
const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

/**
 * Fallback keyword-based moderation
 */
function fallbackModeration(text) {
  console.log("Using fallback keyword-based moderation");
  
  const lowerText = text.toLowerCase();
  const bannedWords = [
    'suicide', 'kill yourself', 'self-harm', 'self harm', 'kys',
    'spam', 'viagra', 'casino', 'scam', 'hate speech'
  ];
  
  const foundWords = bannedWords.filter(word => lowerText.includes(word));
  
  if (foundWords.length > 0) {
    return {
      safety: false,
      categories: foundWords,
      fallback: true
    };
  }
  
  return {
    safety: true,
    categories: [],
    fallback: true
  };
}

/**
 * Moderate content using the LLM API (matches provided sample code)
 * @param {string} text - Content to moderate
 * @returns {Promise<{safety: boolean, categories: array, fallback?: boolean}>}
 */
export const moderateText = async (text) => {
  if (!text || typeof text !== 'string' || text.trim().length === 0) {
    throw new Error("Invalid text for moderation");
  }

  console.log("🔍 Moderating content:", text.substring(0, 50) + "...");

  // Try to connect to the moderation API with retries
  for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
    try {
      // Wait to respect rate limits
      await waitForRateLimit();

      // Call the LLM API
      const response = await axios.post(
        MODERATION_API_URL,
        { prompt: text },
        {
          timeout: MODERATION_TIMEOUT_MS,
          headers: { "Content-Type": "application/json" }
        }
      );

      // Expected response format: { safety: true/false, categories: [...] }
      const { safety, categories } = response.data;

      console.log("Moderation result:", { safety, categories });

      return {
        safety: safety === true || safety === "safe" || safety === "Safe",
        categories: categories || [],
        fallback: false
      };

    } catch (error) {
      console.error(`Moderation attempt ${attempt}/${MAX_RETRIES} failed:`, error.message);

      // Handle specific error cases
      if (error.response?.status === 429) {
        console.error("Rate limit exceeded (429)");
        if (attempt < MAX_RETRIES) {
          await sleep(RETRY_DELAY * attempt);
          continue;
        }
        throw new Error("Rate limit exceeded. Please wait a moment and try again.");
      }
      
      if (error.response?.status === 503 || error.response?.status === 502) {
        console.error("Moderation service unavailable (503/502)");
        if (attempt < MAX_RETRIES) {
          await sleep(RETRY_DELAY * attempt);
          continue;
        }
        throw new Error("Moderation service is temporarily unavailable. Please try again later.");
      }

      if (error.code === 'ECONNABORTED' || error.code === 'ETIMEDOUT') {
        console.error("Moderation request timed out");
        if (attempt < MAX_RETRIES) {
          await sleep(RETRY_DELAY * attempt);
          continue;
        }
        throw new Error("Moderation request timed out. Please try again.");
      }

      // Handle DNS resolution failures (EAI_AGAIN, ENOTFOUND)
      if (error.code === 'EAI_AGAIN' || error.code === 'ENOTFOUND') {
        console.error(" Cannot resolve moderation service hostname:", error.code);
        console.error("   Check if 'cmsai' is in your /etc/hosts file or use 'localhost'");
        
        if (attempt < MAX_RETRIES) {
          await sleep(RETRY_DELAY * attempt);
          continue;
        }
        
        // Use fallback after all retries exhausted
        console.log("All retry attempts failed - using fallback moderation");
        return fallbackModeration(text);
      }

      //  Handle connection refused
      if (error.code === 'ECONNREFUSED') {
        console.error("Cannot connect to moderation service - connection refused");
        console.error("   Check if the moderation API is running on port 8000");
        
        if (attempt < MAX_RETRIES) {
          await sleep(RETRY_DELAY * attempt);
          continue;
        }
        
        // Use fallback after all retries exhausted
        console.log("All retry attempts failed - using fallback moderation");
        return fallbackModeration(text);
      }

      // Generic network errors
      if (error.code && (error.code.startsWith('E') || error.code === 'ECONNRESET')) {
        console.error(`Network error: ${error.code}`);
        
        if (attempt < MAX_RETRIES) {
          await sleep(RETRY_DELAY * attempt);
          continue;
        }
        
        // Use fallback after all retries exhausted
        console.log("All retry attempts failed - using fallback moderation");
        return fallbackModeration(text);
      }

      // If this is the last attempt, use fallback
      if (attempt === MAX_RETRIES) {
        console.error("All moderation attempts failed - using fallback");
        return fallbackModeration(text);
      }

      // Wait before retry
      await sleep(RETRY_DELAY * attempt);
    }
  }

  // Fallback if all retries failed
  console.log("Moderation service completely unavailable - using fallback");
  return fallbackModeration(text);
};

/**
 * Express middleware for content moderation with better error handling
 */
export const moderateContentMiddleware = async (req, res, next) => {
  try {
    const { text, content, body, title } = req.body;
    const textToModerate = text || content || body || title;

    if (!textToModerate) {
      return res.status(400).json({ 
        error: "No content provided for moderation" 
      });
    }

    const result = await moderateText(textToModerate);

    if (!result.safety) {
      console.log("Content rejected by moderation");
      return res.status(403).json({ 
        error: "Content rejected by moderation filter",
        categories: result.categories,
        message: "Your post could not be published due to safety concerns. Please review community guidelines.",
        usedFallback: result.fallback
      });
    }

    // Attach moderation result to request for later use
    req.moderationResult = result;
    
    // Log if fallback was used
    if (result.fallback) {
      console.log("Content approved using fallback moderation");
    }
    
    next();

  } catch (error) {
    console.error("Moderation middleware error:", error.message);
    
    // Check  rate limit error
    if (error.message.includes("Rate limit")) {
      return res.status(429).json({ 
        error: "Rate limit exceeded",
        message: "Too many requests. Please wait a moment and try again.",
        retryAfter: 2
      });
    }

    // Check if it's a service unavailable error
    if (error.message.includes("unavailable") || error.message.includes("connect")) {
      return res.status(503).json({ 
        error: "Service unavailable",
        message: "Moderation service is temporarily unavailable. Please try again later."
      });
    }

    // Generic error
    res.status(503).json({ 
      error: "Moderation failed",
      message: "Unable to moderate content at this time. Please try again later."
    });
  }
};

export default { moderateText, moderateContentMiddleware };

File: backend1/services/moderation_service.py
"""
Moderation service — LLM API + keyword fallback, structured logging.
"""
from __future__ import annotations

import time

import requests

from config import config
from utils.logger import get_logger

logger = get_logger(__name__)

MODERATION_API_URL = config.MODERATION_API_URL
MODERATION_TIMEOUT_S = config.MODERATION_TIMEOUT_MS / 1000
MAX_RETRIES = 2
RETRY_DELAY_S = 1
_last_request_time = 0.0
MIN_REQUEST_INTERVAL_S = 2.0


def _wait_for_rate_limit():
    global _last_request_time
    now = time.time()
    elapsed = now - _last_request_time
    if elapsed < MIN_REQUEST_INTERVAL_S:
        wait_time = MIN_REQUEST_INTERVAL_S - elapsed
        logger.info("Moderation rate-limit", extra={"wait_seconds": round(wait_time, 2)})
        time.sleep(wait_time)
    _last_request_time = time.time()


def _fallback_moderation(text):
    logger.info("Using fallback keyword-based moderation")
    lower_text = text.lower()
    banned_words = [
        "suicide", "kill yourself", "self-harm", "self harm", "kys",
        "spam", "viagra", "casino", "scam", "hate speech",
    ]
    found_words = [w for w in banned_words if w in lower_text]
    if found_words:
        return {"safety": False, "categories": found_words, "fallback": True}
    return {"safety": True, "categories": [], "fallback": True}


def moderate_text(text):
    """
    Moderate content using the external LLM moderation API, with retries
    and a keyword-based fallback if the service is unreachable.
    Returns: {"safety": bool, "categories": list, "fallback": bool}
    """
    if not text or not isinstance(text, str) or not text.strip():
        raise ValueError("Invalid text for moderation")

    logger.info("Moderating content", extra={"text_preview": text[:50]})

    for attempt in range(1, MAX_RETRIES + 1):
        try:
            _wait_for_rate_limit()
            response = requests.post(
                MODERATION_API_URL,
                json={"prompt": text},
                timeout=MODERATION_TIMEOUT_S,
                headers={"Content-Type": "application/json"},
            )
            response.raise_for_status()
            data = response.json()
            safety = data.get("safety")
            categories = data.get("categories") or []
            logger.info("Moderation result", extra={
                "safety": safety, "attempt": attempt, "fallback": False
            })
            return {
                "safety": safety is True or safety == "safe" or safety == "Safe",
                "categories": categories,
                "fallback": False,
            }
        except requests.exceptions.Timeout:
            logger.warning("Moderation timed out", extra={"attempt": attempt})
            if attempt < MAX_RETRIES:
                time.sleep(RETRY_DELAY_S * attempt)
                continue
            raise Exception("Moderation request timed out. Please try again.")
        except requests.exceptions.HTTPError as err:
            status = err.response.status_code if err.response is not None else None
            logger.warning("Moderation HTTP error", extra={"attempt": attempt, "status": status})
            if status == 429 and attempt < MAX_RETRIES:
                time.sleep(RETRY_DELAY_S * attempt)
                continue
            if status in (502, 503) and attempt < MAX_RETRIES:
                time.sleep(RETRY_DELAY_S * attempt)
                continue
            if attempt == MAX_RETRIES:
                logger.warning("All moderation attempts failed - using fallback")
                return _fallback_moderation(text)
            time.sleep(RETRY_DELAY_S * attempt)
        except (requests.exceptions.ConnectionError, requests.exceptions.RequestException) as err:
            logger.warning("Moderation connection error", extra={"attempt": attempt, "error": str(err)})
            if attempt < MAX_RETRIES:
                time.sleep(RETRY_DELAY_S * attempt)
                continue
            logger.warning("All retry attempts failed - using fallback moderation")
            return _fallback_moderation(text)

    logger.warning("Moderation service unavailable - using fallback")
    return _fallback_moderation(text)